### DISCUSSION TITLE
How to get automated updates with renovate?

### DISCUSSION THREAD

**Discussion by internalsystemerror** [nrwl/nx#10003] [Show and Tell]
If using renovate to automatically update dependencies, I need some way of trying to run the nx migrate command instead. Do you have any suggested methods for this? I've been trying to use a [postUpgradeTask](https://docs.renovatebot.com/configuration-options/#postupgradetasks), however this occurs after the versions have been bumped, so no migrations are run.

I continually have to do these manually, just in case there is a migration. Does anyone have any suggestions?

**Comment by wrslatz** [comment]
I'll share how we've set this up when I'm back at my computer (need to not forget)

**Comment by edbzn** [comment]
Hi, in case you are using GitHub actions we wrote an action for that:  https://github.com/jscutlery/semver/blob/main/.github/workflows/nx-migrate.yml

**Comment by wrslatz** [comment]
Here's how we've setup Renovate [postUpgradeTasks](https://docs.renovatebot.com/configuration-options/#postupgradetasks) to automate Nx upgrades and migrations. Pieced these together through trial and error and finding some examples online (I'll link to them if I can find them).

A couple of prerequisites:
1. This requires self hosted Renovate where `postUpgradeTasks` are allowed
2. This requires [allowPostUpgradeCommandTemplating](https://docs.renovatebot.com/self-hosted-configuration/#allowpostupgradecommandtemplating) enabled and [allowedPostUpgradeCommands](https://docs.renovatebot.com/self-hosted-configuration/#allowedpostupgradecommands) setup with the correct allowed commands (I don't happen to have this handy for the Nx commands)
3. I have not updated this for Nx 14 where the `@nrwl/workspace` command has changed. I can update this with more info once we get there.

The best setup I've found is to extend the normal Nrwl group from Renovate, apply and run migrations explicitly from the prior version to the new version. This will show all of the Nrwl packages updated by Renovate together

`.whitesource`
```json
  {
    "extends": ["config:base"],
    "packageRules": [
      {
        "extends": "monorepo:nrwl",
        "groupName": "Nrwl Nx monorepo",
        "postUpgradeTasks": {
          "commands": [
            "npx nx migrate @nrwl/workspace@{{{newVersion}}} --from=@nrwl/workspace@{{{currentVersion}}}",
            "npm install && [ -f migrations.json ] && npx nx migrate --run-migrations=migrations.json"
          ],
          "fileFilters": [
            "**/**"
          ],
          "executionMode": "branch"
        }
      }
    ]
  }
```

**Reply by wrslatz** [reply]
If you need to troubleshoot the templating, I was able to use `prBodyNotes` option inside the package rule there to troubleshoot
```
        "prBodyNotes": [
          "Command: npx nx migrate @nrwl/workspace@{{{newVersion}}} --from=\"@nrwl/workspace@{{{currentVersion}}}\""
        ]
```

**Reply by wrslatz** [reply]
If that doesn't work, you can try the previous method I tried. I didn't like this method as much since it depending on disabling presets from Renovate and only showed `@nrwl/workspace` in the Renovate config
```
  {
    "extends": ["config:base"],
    "ignorePresets": ["group:nrwlMonorepo","monorepo:nrwl"],
    "packageRules": [
      {
        "description": "Disable Renovate grouping of Nrwl monorepo updates",
        "matchSourceUrlPrefixes": [
          "https://github.com/nrwl/"
        ],
        "excludePackageNames": [
          "@nrwl/workspace"
        ],
        "matchUpdateTypes": [
          "major",
          "minor",
          "patch"
        ],
        "enabled": false
      },
      {
        "description": "Automate Nrwl Nx updates with automated migrations",
        "matchPackageNames": [
          "@nrwl/workspace"
        ],
        "groupName": "Nrwl Nx",
        "commitMessageTopic": "Nrwl Nx",
        "postUpgradeTasks": {
          "commands": [
            "npm install --save-dev @nrwl/workspace@{{{currentVersion}}}",
            "npx nx migrate {{{newVersion}}}",
            "npm install && [ -f migrations.json ] && npx nx migrate --run-migrations=migrations.json"
          ],
          "fileFilters": [
            "**/**"
          ]
        }
      }
    ]
  }
```

**Reply by internalsystemerror** [reply]
Thanks for these examples, they help a lot!

**Comment by d-koppenhagen** [comment]
Meanwhile `npx nx migrate --run-migrations` seems to already proceed an `npm install`, so no extra command needed.
Also Nx introduced the `ifExists` option which let's you not fail if the migrations file does not exist:

```bash
npx nx migrate @nrwl/workspace@{{{newVersion}}} --from=@nrwl/workspace@{{{currentVersion}}}
npx nx migrate --run-migrations=migrations.json --create-commits --ifExists
```
